File: cs/days/day04.cs
namespace Shunty.AoC.Days;

// https://adventofcode.com/2024/day/4 - Ceres Search

public class Day04 : AocDaySolver
{
    public int DayNumber => 4;
    public string Title => "Ceres Search";

    public async Task Solve()
    {
        var input = (await File.ReadAllLinesAsync(AocUtils.FindInputFile(DayNumber))).ToImmutableList();
        //var input = TestData.Split('\n').ToList();

        int part1 = 0, part2 = 0;
        for (int y = 0; y < input.Count; y++)
        {
            var row = input[y];
            for (int x = 0; x < row.Length; x++)
            {
                if ('X' == row[x])
                {
                    var mas = FindMAS(x, y, input);
                    part1 += mas;
                }
                else if ('A' == row[x])
                {
                    if (FindX_MAS(x, y, input))
                        part2 += 1;
                }
            }
        }
        this.ShowDayResult(1, part1);
        this.ShowDayResult(2, part2);
    }

    private static readonly (int dx, int dy)[] Directions = [(0,-1), (1,-1), (1,0), (1,1), (0,1), (-1,1), (-1,0), (-1,-1)];
    private static int FindMAS(int X, int Y, IReadOnlyList<string> source)
    {
        var result = 0;
        foreach (var (dx, dy) in Directions)
        {
            int px = X + dx;
            int py = Y + dy;
            if (source.CharAt(px, py) == 'M'
              && source.CharAt(px+dx, py+dy) == 'A'
              && source.CharAt(px+dx+dx, py+dy+dy) == 'S')
            {
                result += 1;
            }
        }
        return result;
    }

    private static bool FindX_MAS(int X, int Y, IReadOnlyList<string> source)
    {
        // pt is an 'A'. We need to find M and S in both diagonals
        // ie ((above-left == S && below-right == M) || (above-left == M && below-right == S))
        //    && ((above-right == S && below-left == M) || (above-right == M && below-left == S))

        var upL = source.CharAt(X-1, Y-1);
        var upR = source.CharAt(X+1, Y-1);
        var dnL = source.CharAt(X-1, Y+1);
        var dnR = source.CharAt(X+1, Y+1);

        return ((upL == 'S' && dnR == 'M') || (upL == 'M' && dnR == 'S'))
            && ((upR == 'S' && dnL == 'M') || (upR == 'M' && dnL == 'S'));
    }

    // Test data: P1 expect 18; P2 expect 9;
    private const string TestData = """ 
    MMMSXXMASM
    MSAMXMSMSA
    AMXSXMAAMM
    MSAMASMSMX
    XMASAMXAMM
    XXAMMXXAMA
    SMSMSASXSS
    SAXAMASAAA
    MAMMMXMMMM
    MXMXAXMASX
    """;
}

File: rs/src/days/day04.rs
use crate::aocutils::*;

pub fn run() {
    let day = 4;
    let filename = find_input_file(&day).unwrap();
    let lines = get_lines_from_file(&filename).unwrap();
    let chars = lines.into_iter().map(|line| line.chars().collect::<Vec<char>>()).collect::<Vec<Vec<char>>>();

    let mut part1 = 0;
    let mut part2 = 0;
    for y in 0..chars.len() {
        for x in 0..chars[y].len() {
            let ch = chars[y][x];
            if ch == 'X' {
                part1 += find_mas(&x, &y, &chars);
            } else if ch == 'A' {
                part2 += find_x_mas(&x, &y, &chars);
            }
        }
    }
    print_day_result(&1, part1);
    print_day_result(&2, part2);
}

fn find_mas(x: &usize, y: &usize, grid: &[Vec<char>]) -> i32 {
    let mut result = 0;
    let dirs = vec![(0, -1), (1, -1), (1, 0), (1, 1), (0, 1), (-1, 1), (-1, 0), (-1, -1)];
    for (dx,dy) in dirs {
        let mut px = *x as i32 + dx;
        let mut py = *y as i32 + dy;
        if px >= 0 && px < grid[0].len() as i32 && py >= 0 && py < grid.len() as i32 && grid[py as usize][px as usize] == 'M' {
            px += dx;
            py += dy;
            if px >= 0 && px < grid[0].len() as i32 && py >= 0 && py < grid.len() as i32 && grid[py as usize][px as usize] == 'A' {
                px += dx;
                py += dy;
                if px >= 0 && px < grid[0].len() as i32 && py >= 0 && py < grid.len() as i32 && grid[py as usize][px as usize] == 'S' {
                    result += 1;
                }
            }
        }
    }
    result
}

fn char_at(x: &i32, y: &i32, grid: &[Vec<char>]) -> char {
    if *x < 0 || *x >= grid[0].len() as i32 || *y < 0 || *y >= grid.len() as i32 {
        return ' ';
    }
    grid[*y as usize][*x as usize]
}

fn find_x_mas(x: &usize, y: &usize, grid: &[Vec<char>]) -> i32 {
    let upl = char_at(&(*x as i32 - 1), &(*y as i32 - 1), grid);
    let upr = char_at(&(*x as i32 + 1), &(*y as i32 - 1), grid);
    let dnl = char_at(&(*x as i32 - 1), &(*y as i32 + 1), grid);
    let dnr = char_at(&(*x as i32 + 1), &(*y as i32 + 1), grid);

    if (upl == 'M' && dnr == 'S' || upl == 'S' && dnr == 'M')
    && (upr == 'M' && dnl == 'S' || upr == 'S' && dnl == 'M') {
        return 1;
    }
    0
}